src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_consumer_power_formula.py

```diff
@@ -50,95 +47,140 @@ def generate(self) -> FormulaEngine[Power]:
 
-        builder.push_component_metric(grid_meter.component_id, nones_are_zeros=False)
+        def non_consumer_component(component: Component) -> bool:
+            """
+            Check if a component is not a consumer component.
+
+            Args:
+                component: The component to check.
 
-        for successor in successors:
+            Returns:
+                True if the component is not a consumer component, False otherwise.
+            """
             # If the component graph supports additional types of grid successors in the
             # future, additional checks need to be added here.
-            if (
-                component_graph.is_battery_chain(successor)
-                or component_graph.is_chp_chain(successor)
-                or component_graph.is_pv_chain(successor)
-                or component_graph.is_ev_charger_chain(successor)
-            ):
+            return (
+                component_graph.is_battery_chain(component)
+                or component_graph.is_chp_chain(component)
+                or component_graph.is_pv_chain(component)
+                or component_graph.is_ev_charger_chain(component)
+            )
+
+        # join all non consumer components reachable from the different grid meters
+        non_consumer_components: set[Component] = set()
+        for grid_meter in grid_meters:
+            non_consumer_components = non_consumer_components.union(
+                component_graph.dfs(grid_meter, set(), non_consumer_component)
+            )
+
+        # push all grid meters
+        for idx, grid_meter in enumerate(grid_meters):
+            if idx > 0:
                 builder.push_oper("-")
-                nones_are_zeros = True
-                if successor.category == ComponentCategory.METER:
-                    nones_are_zeros = False
-                builder.push_component_metric(
-                    successor.component_id, nones_are_zeros=nones_are_zeros
-                )
+            builder.push_component_metric(
+                grid_meter.component_id, nones_are_zeros=False
+            )
+
+        # push all non consumer components and subtract them from the grid meters
+        for component in non_consumer_components:
+            builder.push_oper("-")
+            builder.push_component_metric(
+                component.component_id,
+                nones_are_zeros=component.category != ComponentCategory.METER,
+            )
 
         return builder.build()
 
     def _gen_without_grid_meter(
         self,
         builder: ResampledFormulaBuilder[Power],
-        grid_successors: abc.Iterable[Component],
+        grid: Component,
     ) -> FormulaEngine[Power]:
         """Generate formula for calculating consumer power without a grid meter.
 
         Args:
             builder: The formula engine builder.
-            grid_successors: The grid successors.
+            grid: The grid successors.
```

Comment:
you might also want to amend the documentation here

Reply:
Well spotted. Thanks!